python/ray/_private/arrow_serialization.py

```diff
@@ -226,195 +312,346 @@ def _copy_array_if_needed(a: "pyarrow.Array") -> "pyarrow.Array":
+    )
 
-    if pa.types.is_null(a.type):
-        # Return NullArray as is.
-        return a
 
+def _null_array_to_array_payload(a: "pyarrow.NullArray") -> "PicklableArrayPayload":
+    """Serialize null array to PicklableArrayPayload."""
+    # Buffer scheme: [None]
+    return PicklableArrayPayload(
+        type=a.type,
+        length=len(a),
+        buffers=[None],  # Single null buffer is expected.
+        null_count=a.null_count,
+        offset=0,
+        children=[],
+    )
+
+
+def _primitive_array_to_array_payload(a: "pyarrow.Array") -> "PicklableArrayPayload":
+    """Serialize primitive (numeric, temporal, boolean) arrays to
+    PicklableArrayPayload.
+    """
+    assert _is_primitive(a.type), a.type
+    # Buffer scheme: [bitmap, data]
     buffers = a.buffers()
-    bitmap = buffers[0]
-    buf = buffers[1]
-    # Let remaining buffers be handled downstream.
-    buffers = buffers[2:]
-    children = None
-    if pa.types.is_struct(a.type) or pa.types.is_union(a.type):
-        # Struct and union arrays directly expose children arrays, which are easier
-        # to work with than the raw buffers.
-        children = [a.field(i) for i in range(a.type.num_fields)]
-        buffers = None
-    if pa.types.is_map(a.type):
-        if isinstance(a, pa.lib.ListArray):
-            # Map arrays directly expose the one child array in pyarrow>=7.0.0, which
-            # is easier to work with than the raw buffers.
-            children = [a.values]
-            buffers = None
-        else:
-            # In pyarrow<7.0.0, the child array is not exposed, so we work with the key
-            # and item arrays.
-            if bitmap is not None:
-                bitmap = _copy_bitpacked_buffer_if_needed(bitmap, a.offset, len(a))
-            offset_buf, data_offset, data_length = _copy_offsets_buffer_if_needed(
-                buf, a.type, a.offset, len(a)
-            )
-            offsets = pa.Array.from_buffers(
-                pa.int32(), len(a) + 1, [bitmap, offset_buf]
-            )
-            keys = _copy_array_if_needed(a.keys.slice(data_offset, data_length))
-            items = _copy_array_if_needed(a.items.slice(data_offset, data_length))
-            return pa.MapArray.from_arrays(offsets, keys, items)
-    return _copy_array_buffers_if_needed(
-        bitmap,
-        buf,
-        a.type,
-        a.offset,
-        len(a),
-        a.null_count,
-        buffers=buffers,
+    assert len(buffers) <= 2, len(buffers)
+
+    # Copy bitmap buffer, if needed.
+    if a.null_count > 0:
+        bitmap_buf = _copy_bitpacked_buffer_if_needed(buffers[0], a.offset, len(a))
+    else:
+        bitmap_buf = None
+
+    # Copy data buffer, if needed.
+    if len(buffers) > 1:
+        assert len(buffers) == 2, len(buffers)
+        data_buf = _copy_buffer_if_needed(buffers[1], a.type, a.offset, len(a))
+    else:
+        assert len(buffers) == 1, len(buffers)
+        data_buf = None
```

Comment:
Correct, although I don't that this can happen. We could probably be stricter with our assertions.

I think some of this might be a carry over of some shared logic between primitive arrays and other array types, so I think that this handling might be too permissive... E.g. I think we can guarantee that `len(buffers) == 2` always holds for primitive arrays.

Let me make that change and see if everything passes!